### PR TITLE
Node pools

### PR DESCRIPTION
**Pull request by dutsmiller**
- manage additional node pools within module

### CODE REVIEW

File: example/node-pools/main.tf [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      source  = "hashicorp/azurerm"
+      version = "=2.48.0"
+    }
+    random = {
+      source  = "hashicorp/random"
+      version = "=2.3.0"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = "=1.13.0"
```
Comment (hiscjo01) [inline example/node-pools/main.tf:13]: Should we be using version 2 of the Kubernetes provider now its gone GA?

File: example/node-pools/main.tf [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    }
+    helm = {
+      source  = "hashicorp/helm"
+      version = "=1.3.2"
+    }
+  }
+  required_version = "=0.14.7"
```
Comment (hiscjo01) [inline example/node-pools/main.tf:20]: This should be >=

File: example/node-pools/main.tf [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+provider "azurerm" {
+  features {}
+}
+
+provider "kubernetes" {
+  host                   = module.kubernetes.host
+  client_certificate     = base64decode(module.kubernetes.client_certificate)
+  client_key             = base64decode(module.kubernetes.client_key)
+  cluster_ca_certificate = base64decode(module.kubernetes.cluster_ca_certificate)
+  load_config_file       = false
```
Comment (hiscjo01) [inline example/node-pools/main.tf:32]: If you upgrade to V2 of the provider this line won't be needed

File: example/node-pools/main.tf [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      source  = "hashicorp/random"
+      version = "=2.3.0"
+    }
+    kubernetes = {
+      source  = "hashicorp/kubernetes"
+      version = "=1.13.0"
+    }
+    helm = {
+      source  = "hashicorp/helm"
+      version = "=1.3.2"
```
Comment (hiscjo01) [inline example/node-pools/main.tf:17]: V2 of the Helm provider has also gone GA

File: example/node-pools/main.tf [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  source = "github.com/Azure-Terraform/terraform-azurerm-subscription-data.git?ref=v1.0.0"
+  subscription_id = data.azurerm_subscription.current.subscription_id
+}
+
+module "naming" {
+  source = "github.com/Azure-Terraform/example-naming-template.git?ref=v1.0.0"
+}
+
+module "metadata" {
+  source = "github.com/Azure-Terraform/terraform-azurerm-metadata.git?ref=v1.1.0"
```
Comment (hiscjo01) [inline example/node-pools/main.tf:72]: We have a v1.5.0 now
Not sure if it has anything of use for you - but might be worth updating if you're already editing the file

File: main.tf [hunk 6]
```diff
@@ -40,28 +40,45 @@ module "subnet_config" {
 
 resource "azurerm_kubernetes_cluster" "aks" {
   depends_on          = [azurerm_role_assignment.route_table_network_contributor]
+  #lifecycle {
+  #  ignore_changes = [(local.node_pools[var.default_node_pool].enable_auto_scaling ? "node_count" : "")] 
+  #}
```
Comment (hiscjo01) [inline main.tf:45]: Should this be removed instead of commented out?

File: main.tf [hunk 7]
```diff
@@ -40,28 +40,45 @@ module "subnet_config" {
...
-    max_count           = (var.default_node_pool_enable_auto_scaling ? var.default_node_pool_node_max_count : null)
-    availability_zones  = var.default_node_pool_availability_zones
-    vnet_subnet_id      = (var.aks_managed_vnet ? null : var.node_pool_subnets[var.default_node_pool_subnet].id)
-    tags                = var.tags
+    name                         = local.node_pools[var.default_node_pool].name
+    vm_size                      = local.node_pools[var.default_node_pool].vm_size
+    os_disk_size_gb              = local.node_pools[var.default_node_pool].os_disk_size_gb
+    os_disk_type                 = local.node_pools[var.default_node_pool].os_disk_type
+    availability_zones           = local.node_pools[var.default_node_pool].availability_zones
+    enable_auto_scaling          = local.node_pools[var.default_node_pool].enable_auto_scaling
```
Comment (hiscjo01) [inline main.tf:65]: Would it be valuable to add defaults to any of theses?

File: main.tf [hunk 8]
```diff
@@ -97,6 +114,41 @@ resource "azurerm_kubernetes_cluster" "aks" {
   }
 }
 
+resource "azurerm_kubernetes_cluster_node_pool" "example" {
```
Comment (hiscjo01) [inline main.tf:117]: Resource needs to be renamed from "example"

**Review by hiscjo01 — CHANGES_REQUESTED** [review]

File: example/node-pools/main.tf [hunk 9]
```diff
@@ -133,16 +133,24 @@ module "kubernetes" {
       name   = "system"
       subnet = "private"
     }
-    web = {
-      name                = "web"
+    llinux_web = {
```
Comment (fabiendelpierre) [inline example/node-pools/main.tf:136]: Typo here, `llinux`.

File: subnet_config/variables.tf [hunk 10]
```diff
@@ -1,15 +1,15 @@
 variable "principal_id" {
-  description = "Id of principal which manages AKS"
+  description = "Principal ID which manages AKS."
   type        = string
 }
 
 variable "subnet_info" {
-  description = "Azure virtual network subnet id."
-  type        = map(object({
+  description = "Azure subnet info."
```
Comment (fabiendelpierre) [inline subnet_config/variables.tf:7]: Any way to make this more descriptive?

**Review by fabiendelpierre — CHANGES_REQUESTED** [review]

File: example/node-pools/main.tf (hunk 5 above)
Comment (dutsmiller) [reply example/node-pools/main.tf]: good catch

File: example/node-pools/main.tf (hunk 9 above)
Comment (dutsmiller) [reply example/node-pools/main.tf]: fixed

File: main.tf (hunk 8 above)
Comment (dutsmiller) [reply main.tf:114]: good catch

File: example/node-pools/main.tf (hunk 1 above)
Comment (dutsmiller) [reply example/node-pools/main.tf]: we actually didn't need it, removed.

File: example/node-pools/main.tf (hunk 2 above)
Comment (dutsmiller) [reply example/node-pools/main.tf]: fixed

File: main.tf (hunk 7 above)
Comment (dutsmiller) [reply main.tf:53]: they automatically inherit the values in the variable node_pool_defaults

**Pull request merged**